File: c++ Data Structure/binary-tree.cpp
#include<iostream>
using namespace std;
#define NULL 0
struct tree
{
	int data;
	struct tree *left;
	struct tree *right;
};
typedef struct tree node;
node *root=NULL;
void display(node *root,node *prev,int height,char position)
{
	if(root!=NULL)
	{
		if(height==0)	
		{
			cout<<"root node: "<<root->data<<endl;
		}
		else
		{
			cout<<"level: "<<height<<" node value: "<<root->data<<" position: "<<position<<" of "<<prev->data<<endl;
		}
		display(root->left,root,height+1,'L');
		display(root->right,root,height+1,'R');
	}
}
int main()
{
	int ch=0;
	cout<<"BUG: DELETION OF NODES IS NOT ALWAYS CORRECT. DELETION OF ROOT NODE OR NODE WITH ONE LEAF NODE WORKS!"<<endl;
	while(ch<5)
	{
		cout<<"!.insert 2.delete 3.display"<<endl;
		cin>>ch;
		switch(ch)
		{
			case 1:
			{
				node *newnode;
				newnode=(node *)malloc(sizeof(node));
				cout<<"enter data";
				cin>>newnode->data;
				newnode->left=NULL;
				newnode->right=NULL;
				if(root==NULL)
				{
					root=newnode;
				}
				else 
				{
					node *temp;
					temp=root;
					while(1)
					{
						if(temp->data < newnode->data)
						{
							if(temp->right==NULL)
							{
								temp->right=newnode;
								break;
							}
							temp=temp->right;
						}
						else
						{
							if(temp->left==NULL)
							{
								temp->left=newnode;
								break;
							}
							temp=temp->left;
						}
					}
				}
				break;
			}
			case 2:
			{
				int ele;
				node *temp,*pleft,*pright;
				temp=root;
				cout<<"enter node value to be deleted. WARNING: Do not delete the root node"<<endl;
				cin>>ele;
				while(1)
				{
					if(temp->data < ele)
					{
						if(temp->right->data == ele)
						{
							pright=temp->right->right;
							pleft=temp->right->left;
							free(temp->right);
							temp->right=pright;
							temp->left=pleft;
							break;
						}
						else
						{
							temp=temp->right;
						}
					}
					else
					{
						if(temp->left->data == ele)
						{
							pright=temp->left->right;
							pleft=temp->left->left;
							free(temp->left);
							temp->right=pright;
							temp->left=pleft;
							break;
						}
						else
						{
							temp=temp->left;
						}
					}
				}
				break;
			}
			case 3:
			{
				node *temp;
				temp=root;
				cout<<"Displaying using depth first search"<<endl;
				display(root,root,0,'L');
				break;
			}
		}
	}
}

/* OUTPUT: 
!.insert 2.delete 3.display
1
enter data10
!.insert 2.delete 3.display
1
enter data20
!.insert 2.delete 3.display
1
enter data5
!.insert 2.delete 3.display
3
Displaying using depth first search
root node: 10
level: 1 node value: 5 position: L of 10
level: 1 node value: 20 position: R of 10
!.insert 2.delete 3.display
2
enter node value to be deleted. WARNING: Do not delete the root node
20
!.insert 2.delete 3.display
3
Displaying using depth first search
root node: 10
!.insert 2.delete 3.display
5
*/ 

File: Python Data Structure/binary-tree.py
class node:
	def __init__(self,data):
		self.data=data
		self.left=None
		self.right=None
	def insert(self,data):
		if(self.data < data):
			if(self.right==None):
				self.right=node(data)
			else:
				self.right.insert(data)
		else:
			if(self.left==None):
				self.left=node(data)
			else:
				self.left.insert(data)
	def display(self):
		if(self.left is not None):
			self.left.display()
		print(self.data)
		if(self.right is not None):
			self.right.display()
	def search(self,data):
		global elementfound
		if(self.data==data):
			elementfound=True
			pass
		else:
			if(self.data < data):
				if(self.right is not None):
					self.right.search(data)
			else:
				if(self.left is not None):
					self.left.search(data)

elementfound=False
print("enter root value")
obj=node(int(input()))
ch=0
while(ch<4):
	print("1.insert 2.display 3.search")
	ch=int(input())
	if(ch==1):
		print("Enter value to insert")
		element=int(input())
		obj.insert(element)
	elif(ch==2):
		print("Depth Traversal")
		obj.display()
	elif(ch==3):
		elementfound=False
		print("Enter element to search")
		element=int(input())
		obj.search(element)
		if(elementfound):
			print("element found")
		else:
			print("element not found")
	else:
		break

